<a href="https://colab.research.google.com/github/Eshachatterjee01/data_science_all/blob/main/PCA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# PCA


Digit REcognizer i have done in kaggle notebook.because i took dataset from kaggle..

link-- https://www.kaggle.com/code/eshachatterjee01/pca-digit-recognize/edit

**PCA BY Sikhar Sir**

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import StandardScaler,LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score,roc_auc_score

So first we will find the accuracy score without doing any transformation.then we will find accuracy score after applying PCA transformation.

**Without applying PCA transformation**

In [ ]:
data = pd.read_csv("/content/drive/MyDrive/DataScience_and_MachineLearning_Intellipaat/Datasets/nba_final.csv")

In [ ]:
data

,Rk,Player.x,Player_ID,Pos1,Pos2,Age,Tm,G,GS,MP,...,Conference,Role,Fvot,FRank,Pvot,PRank,Mvot,MRank,Score,Play
0,170,A.J. Hammons,hammoaj01,C,NaN,24,DAL,22,0,7.4,...,West,Front,786,123,NaN,NaN,NaN,NaN,83.5,No
1,58,Aaron Brooks,brookaa01,PG,NaN,32,IND,65,0,13.8,...,Est,Back,2474,64,NaN,NaN,NaN,NaN,48.2,No
2,157,Aaron Gordon,gordoaa01,SF,NaN,21,ORL,80,72,28.7,...,Est,Front,22774,29,NaN,NaN,NaN,NaN,40.0,No
3,352,Adreian Payne,paynead01,PF,NaN,25,MIN,18,0,7.5,...,West,Front,861,120,1.0,52.0,NaN,NaN,75.5,No
4,10,Al-Farouq Aminu,aminual01,PF,NaN,26,POR,61,25,29.1,...,West,Front,4971,69,7.0,23.0,NaN,NaN,42.8,No
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1403,109,Zach Collins,colliza01,C,NaN,21,POR,77,0,17.6,...,West,Front,16287,74,1.0,52.0,0.0,8.0,86.8,No
1404,294,Zach LaVine,lavinza01,SG,NaN,23,CHI,63,62,34.5,...,Est,Back,519746,7,17.0,8.0,0.0,8.0,36.5,No
1405,308,Zach Lofton,loftoza01,SG,NaN,26,DET,1,0,4.0,...,Est,Back,136,114,0.0,61.0,0.0,8.0,132.2,No
1406,393,Zaza Pachulia,pachuza01,C,NaN,34,DET,68,3,12.9,...,Est,Front,21525,50,2.0,41.0,0.0,7.0,69.8,No


In [ ]:
data.shape

(1408, 45)

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1408 entries, 0 to 1407
Data columns (total 45 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Rk          1408 non-null   int64  
 1   Player.x    1408 non-null   object 
 2   Player_ID   1408 non-null   object 
 3   Pos1        1408 non-null   object 
 4   Pos2        12 non-null     object 
 5   Age         1408 non-null   int64  
 6   Tm          1408 non-null   object 
 7   G           1408 non-null   int64  
 8   GS          1408 non-null   int64  
 9   MP          1408 non-null   float64
 10  FG          1408 non-null   float64
 11  FGA         1408 non-null   float64
 12  FG.         1404 non-null   float64
 13  X3P         1408 non-null   float64
 14  X3PA        1408 non-null   float64
 15  X3P.        1309 non-null   float64
 16  X2P         1408 non-null   float64
 17  X2PA        1408 non-null   float64
 18  X2P.        1393 non-null   float64
 19  eFG.        1404 non-null  

In [ ]:
data.isnull().sum()

,0
Rk,0
Player.x,0
Player_ID,0
Pos1,0
Pos2,1396
Age,0
Tm,0
G,0
GS,0
MP,0


In [ ]:
data.isnull().sum().sum()

np.int64(2891)

In [ ]:
data.isnull().sum(axis = 1) ##total number of missing values across each row

,0
0,6
1,5
2,5
3,3
4,3
...,...
1403,1
1404,1
1405,4
1406,1


In [ ]:
data.isnull().sum(axis = 0) #total number of missing values in each column.

,0
Rk,0
Player.x,0
Player_ID,0
Pos1,0
Pos2,1396
Age,0
Tm,0
G,0
GS,0
MP,0


In [ ]:
#which row has more than 4 missing value?
data[data.isnull().sum(axis = 1)>4]

,Rk,Player.x,Player_ID,Pos1,Pos2,Age,Tm,G,GS,MP,...,Conference,Role,Fvot,FRank,Pvot,PRank,Mvot,MRank,Score,Play
0,170,A.J. Hammons,hammoaj01,C,NaN,24,DAL,22,0,7.4,...,West,Front,786,123,NaN,NaN,NaN,NaN,83.5,No
1,58,Aaron Brooks,brookaa01,PG,NaN,32,IND,65,0,13.8,...,Est,Back,2474,64,NaN,NaN,NaN,NaN,48.2,No
2,157,Aaron Gordon,gordoaa01,SF,NaN,21,ORL,80,72,28.7,...,Est,Front,22774,29,NaN,NaN,NaN,NaN,40.0,No
8,464,Alan Williams,willial03,C,NaN,24,PHO,47,0,15.1,...,West,Front,607,127,NaN,NaN,NaN,NaN,85.5,No
9,65,Alec Burks,burksal01,SG,NaN,25,UTA,42,0,15.5,...,West,Back,1981,72,NaN,NaN,NaN,NaN,53.0,No
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
430,373,Willie Reed,reedwi02,C,NaN,26,MIA,71,5,14.5,...,Est,Front,1062,112,NaN,NaN,NaN,NaN,81.5,No
472,348,Ben Moore,moorebe01,PF,NaN,22,IND,2,0,4.5,...,Est,Front,267,136,0.0,74.0,0.0,8.0,158.5,No
904,307,Tyler Lydon,lydonty01,PF,NaN,21,DEN,1,0,2.0,...,West,Front,850,135,1.0,47.0,0.0,9.0,118.2,No
1089,282,George King,kingge03,SF,NaN,25,PHO,1,0,6.0,...,West,Front,2964,119,0.0,84.0,0.0,8.0,152.0,No


In [ ]:
#i will save which have less than 5 missing value
data = data[data.isnull().sum(axis = 1) <=5]

In [ ]:
data.shape #before filtering shape was (1408,45)

(1371, 45)

In [ ]:
data.isnull().sum().sum() #before it was 2891

np.int64(2659)

In [ ]:
data.isnull().sum() #Pos2 has more null values we will drop that column

,0
Rk,0
Player.x,0
Player_ID,0
Pos1,0
Pos2,1359
Age,0
Tm,0
G,0
GS,0
MP,0


In [ ]:
data.drop(columns='Pos2', inplace = True)

/tmp/ipython-input-2057724361.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data.drop(columns='Pos2', inplace = True)


In [ ]:
data.head(2)

,Rk,Player.x,Player_ID,Pos1,Age,Tm,G,GS,MP,FG,...,Conference,Role,Fvot,FRank,Pvot,PRank,Mvot,MRank,Score,Play
1,58,Aaron Brooks,brookaa01,PG,32,IND,65,0,13.8,1.9,...,Est,Back,2474,64,NaN,NaN,NaN,NaN,48.2,No
2,157,Aaron Gordon,gordoaa01,SF,21,ORL,80,72,28.7,4.9,...,Est,Front,22774,29,NaN,NaN,NaN,NaN,40.0,No


In [ ]:
data = data.fillna(0)

In [ ]:
data.isnull().sum().sum()

np.int64(0)

In [ ]:
data.duplicated().sum()

np.int64(0)

In [ ]:
data.columns

Index(['Rk', 'Player.x', 'Player_ID', 'Pos1', 'Age', 'Tm', 'G', 'GS', 'MP',
       'FG', 'FGA', 'FG.', 'X3P', 'X3PA', 'X3P.', 'X2P', 'X2PA', 'X2P.',
       'eFG.', 'FT', 'FTA', 'FT.', 'ORB', 'DRB', 'TRB', 'AST', 'STL', 'BLK',
       'TOV', 'PF', 'PTS', 'Salary', 'mean_views', 'Season', 'Conference',
       'Role', 'Fvot', 'FRank', 'Pvot', 'PRank', 'Mvot', 'MRank', 'Score',
       'Play'],
      dtype='object')

In [ ]:
#we will delete column player.x and player id..this colimns are note helpfull for machine learning
data.drop(columns = ['Player.x','Player_ID'],inplace = True)

Encoding the categorical columns

In [ ]:
#first see where its object datatype present in which column
data.select_dtypes('object')

,Pos1,Tm,Season,Conference,Role,Play
1,PG,IND,2016-17,Est,Back,No
2,SF,ORL,2016-17,Est,Front,No
3,PF,MIN,2016-17,West,Front,No
4,PF,POR,2016-17,West,Front,No
5,C,BOS,2016-17,Est,Front,No
...,...,...,...,...,...,...
1403,C,POR,2018-19,West,Front,No
1404,SG,CHI,2018-19,Est,Back,No
1405,SG,DET,2018-19,Est,Back,No
1406,C,DET,2018-19,Est,Front,No


In [ ]:
obj_cols = data.select_dtypes('object').columns

In [ ]:
obj_cols

Index(['Pos1', 'Tm', 'Season', 'Conference', 'Role', 'Play'], dtype='object')

In [ ]:
#converting object into numeric
le = LabelEncoder()
for col in obj_cols:
  data[col] = le.fit_transform(data[col])

In [ ]:
data[col].dtype

dtype('int64')

In [ ]:
#check for data imbalance
data['Play'].value_counts() ##heavily imbalance

,count
Play,
0,1298
1,73


Machine Learning Process

In [ ]:
x = data.drop(columns = 'Play')
y = data['Play']

In [ ]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2,random_state=42,stratify=y)

Mean centering/Scalling/Standardization of the data

In [ ]:
scaler = StandardScaler()
x_train_scale = scaler.fit_transform(x_train)
x_test_scale = scaler.transform(x_test)

**Now we are goinh to apply simple Logistic Regression on the data**

In [ ]:
log_reg = LogisticRegression()
log_reg.fit(x_train_scale,y_train)

LogisticRegression()

In [ ]:
y_pred = log_reg.predict(x_test_scale)


In [ ]:
accuracy_score(y_test,y_pred)  ##accuracy_score is unreliable in the case of imbalance data, so we don't trust it

0.9745454545454545

In [ ]:
#instead i will use roc_auc_score
roc_auc_score(y_test,y_pred)

np.float64(0.8294871794871794)

To check for the overfitting

In [ ]:
y_pred_train = log_reg.predict(x_train_scale)

In [ ]:
roc_auc_score(y_train,y_pred_train)

np.float64(0.920968706398246)

**Now we will transform the data with PCA and check the performance of the model using the PCA transformed data**


In [ ]:
from sklearn.decomposition import PCA


we can mention n_component = 20 ,means how many column or principal component you want that.. but here i used 0.95 that means when i am extracting the information of these 42 columns and creating new columns 95% information has to be extracted properly and 5 % we can loose .based on 95% of information extraction pca will decide the column count..

In [ ]:
pca = PCA(n_components=0.95)

In [ ]:
x_train_pca = pca.fit_transform(x_train_scale)
x_test_pca = pca.transform(x_test_scale)

In [ ]:
x_train_pca.shape #from 41 column become 21

(1096, 21)

Apply the logistic Regression on the pca_transformed data

In [ ]:
log_reg2 = LogisticRegression()

In [ ]:
log_reg2.fit(x_train_pca,y_train)

LogisticRegression()

In [ ]:
y_pred_pca = log_reg2.predict(x_test_pca)

In [ ]:
roc_auc_score(y_test,y_pred_pca)

np.float64(0.7628205128205128)

In [ ]:
y_pred_train = log_reg2.predict(x_train_pca)

In [ ]:
roc_auc_score(y_train,y_pred_train)

np.float64(0.9118663211746727)

Disclaimer:-  after doing pca transformation also still overfitting.training accuracy 0.911 and test accuracy is 0.76. actually logistic regression,linear regression,svm all of these are not suitable for imbalance dataset..most suitable algorithms are tree based (DT,RF,GradientBoost,XGBoost)..because linear,logistic they are not able to understand patterns properly they are doing overlearning in this case.